File: problems/13_1d_dp/0647_palindromic_substrings/solution.py
"""
Problem: 0647. Palindromic Substrings
"""


class Solution:
    def countSubstrings(self, s: str) -> int:
        cnt = 0

        def expand(l: int, r: int) -> int:
            c = 0
            while l >= 0 and r < len(s) and s[l] == s[r]:
                c += 1; l -= 1; r += 1
            return c

        for i in range(len(s)):
            cnt += expand(i, i)
            cnt += expand(i, i + 1)
        return cnt


if __name__ == "__main__":
    sol = Solution()
    assert sol.countSubstrings("abc") == 3
    assert sol.countSubstrings("aaa") == 6
    assert sol.countSubstrings("") == 0
    assert sol.countSubstrings("a") == 1
    assert sol.countSubstrings("aaaa") == 10
    assert sol.countSubstrings("abba") == 6
    print("All tests passed!")

File: problems/13_1d_dp/0647_palindromic_substrings/solution.dart
// 0647. Palindromic Substrings
int countSubstrings(String s) {
  var cnt = 0;
  int expand(int l, int r) {
    var c = 0;
    while (l >= 0 && r < s.length && s[l] == s[r]) { c++; l--; r++; }
    return c;
  }
  for (var i = 0; i < s.length; i++) { cnt += expand(i, i); cnt += expand(i, i + 1); }
  return cnt;
}
void main() {
  if (countSubstrings("abc") != 3) throw 'a';
  if (countSubstrings("aaa") != 6) throw 'b';
  if (countSubstrings("") != 0) throw 'c';
  if (countSubstrings("a") != 1) throw 'd';
  if (countSubstrings("aaaa") != 10) throw 'e';
  if (countSubstrings("abba") != 6) throw 'f';
  print('All tests passed!');
}
